### PR TITLE
Feature expanded story release

### PR DESCRIPTION
**Pull request by kostadriano**
Make release story type [#22000](https://central.cm42.io/projects/central-board-v2#story-22000)

CM42 Central (12)

### CODE REVIEW

File: app/assets/javascripts/components/story/CollapsedStory/CollapsedStoryStateActions.js [hunk 1]
```diff
@@ -40,11 +42,20 @@ class CollapsedStoryStateActions extends React.Component {
...
+
+  getState(story) {
+    return isRelease(story.storyType) && !isAccepted(story)
+      ? status.RELEASE
+      : story.state;
   }
 
   render() {
     const { story, onUpdate } = this.props;
+    const state = this.getState(story);
```
Comment (talyssonoc) [inline app/assets/javascripts/components/story/CollapsedStory/CollapsedStoryStateActions.js]: Avoid having the method `getState` in a class that `setState` does something waaaay different.

File: app/assets/javascripts/components/story/CollapsedStory/CollapsedStoryTitle.js [hunk 2]
```diff
@@ -4,12 +4,23 @@ import { storyPropTypesShape } from '../../../models/beta/story';
...
+        ? <abbr
+          className="Story__initials"
+          title={story.ownedByName}
+        >
+          {story.ownedByInitials}
+        </abbr>
+        : null
+    }
+    {
+      story.storyType === 'release'
```
Comment (talyssonoc) [inline app/assets/javascripts/components/story/CollapsedStory/CollapsedStoryTitle.js]: Don't we have an `isRelease` method somewhere?

File: app/assets/javascripts/components/story/CollapsedStory/CollapsedStoryTitle.js [hunk 3]
```diff
@@ -4,12 +4,23 @@ import { storyPropTypesShape } from '../../../models/beta/story';
...
+        ? story.releaseDate
```
Comment (talyssonoc) [inline app/assets/javascripts/components/story/CollapsedStory/CollapsedStoryTitle.js]: We should have an `isReleased` method, what do you think?

File: app/assets/javascripts/components/story/ExpandedStory/ExpandedStoryRelease/index.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  onEdit
+}) =>
+  <Fragment>
+    <ExpandedStoryTitle
+      story={story}
+      titleRef={titleRef}
+      onEdit={(title) => onEdit({ title })}
+    />
+
+    <ExpandedStoryType story={story}
```
Comment (talyssonoc) [inline app/assets/javascripts/components/story/ExpandedStory/ExpandedStoryRelease/index.js]: Can we move `story` prop to a new line?

File: app/assets/javascripts/components/story/ExpandedStory/ExpandedStoryState.js [hunk 5]
```diff
@@ -11,7 +11,7 @@ const ExpandedStoryState = ({ story, onEdit }) =>
     <select
       value={story._editing.state}
       className="form-control input-sm"
-      onChange={(event) => onEdit({ state: event.target.value })}
```
Comment (talyssonoc) [inline app/assets/javascripts/components/story/ExpandedStory/ExpandedStoryState.js]: Why was this necessary?

File: app/assets/javascripts/components/story/ExpandedStory/index.js [hunk 6]
```diff
@@ -68,83 +48,21 @@ export class ExpandedStory extends React.Component {
...
-
-              <ExpandedStoryTask
-                story={story}
-                onToggle={(task, status) => toggleTask(project.id, story, task, status)}
-                onDelete={(taskId) => deleteTask(project.id, story.id, taskId)}
-                onSave={(task) => createTask(project.id, story.id, task)}
-              />
-            </Fragment>
-            : null
+          Story.isRelease(story._editing.storyType)
```
Comment (talyssonoc) [inline app/assets/javascripts/components/story/ExpandedStory/index.js]: Doesn't the `Story.isRelease` method accept a story instance as parameter rather than the `storyType`?

File: app/assets/javascripts/components/story/ExpandedStory/index.js [hunk 7]
```diff
@@ -68,83 +48,21 @@ export class ExpandedStory extends React.Component {
...
+            ? <ExpandedStoryRelease
+              story={story}
+              titleRef={this.titleRef}
+              onEdit={(newAttributes) => editStory(story.id, newAttributes)}
+            />
+            : <ExpandedStoryDefault
+              story={story}
+              titleRef={this.titleRef}
+              onEdit={(newAttributes) => editStory(story.id, newAttributes)}
+              isNew={Story.isNew(story)}
```
Comment (talyssonoc) [inline app/assets/javascripts/components/story/ExpandedStory/index.js]: We should do this inside the `ExpandedStoryDefault` component, this is a derived info from the `story` prop.

File: app/assets/javascripts/models/beta/story.js [hunk 8]
```diff
@@ -66,6 +67,17 @@ export const isStoryNotEstimated = (storyType, estimate) => storyType === 'featu
 
 export const isRelease = (storyType) => storyType === 'release';
 
+export const releaseIsLate = (story) => {
+  if (story.storyType !== storyTypes.RELEASE) {
```
Comment (talyssonoc) [inline app/assets/javascripts/models/beta/story.js]: We should use the `isRelease` here.

File: app/assets/stylesheets/new_board/_story.scss [hunk 9]
```diff
@@ -4,6 +4,13 @@
     border-bottom: 1px solid $story-border-bottom-color;
     border-top: 1px solid $story-border-top-color;
 
+    &--late-release {
+      background-color: rgba($darkred, .5) !important;
```
Comment (talyssonoc) [inline app/assets/stylesheets/new_board/_story.scss]: Do we need these `!important`?

File: config/locales/es.yml [hunk 10]
```diff
@@ -92,6 +92,7 @@ es:
       accept: "aceptar"
       reject: "rechazar"
       restart: "reanudar"
+      release: "release"
```
Comment (talyssonoc) [inline config/locales/es.yml]: Is it "release" in Spanish?

File: spec/javascripts/components/story/expanded_story/expanded_story_release/expanded_story_release_spec.js [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+  });
+
+  it('renders children components', () => {
+    const wrapper = shallow(
+      <ExpandedStoryRelase
+        {...defaultProps()}
+      />
+    );
+
+    expect(wrapper.find('ExpandedStoryTitle')).toExist();
```
Comment (talyssonoc) [inline spec/javascripts/components/story/expanded_story/expanded_story_release/expanded_story_release_spec.js]: Do we need these tests?

File: spec/javascripts/components/story/expanded_story/expanded_story_spec.js [hunk 12]
```diff
@@ -2,67 +2,79 @@ import React from 'react';
...
+        />,
+        { disableLifecycleMethods: true }
+      );
+
+      expect(wrapper.find(ExpandedStoryDefault)).toExist();
+      expect(wrapper.find(ExpandedStoryRelease)).not.toExist();
+    });
   });
 
   it('adds loading overlay when updating a story', () => {
```
Comment (talyssonoc) [inline spec/javascripts/components/story/expanded_story/expanded_story_spec.js]: Are we checking if it really adds the overlay here? Or is it checking if it adds the overlay _class_?

File: app/assets/stylesheets/new_board/_story.scss (hunk 9 above)
Comment (kostadriano) [reply app/assets/stylesheets/new_board/_story.scss]: yes

File: config/locales/es.yml (hunk 10 above)
Comment (kostadriano) [reply config/locales/es.yml]: `lanzamiento` don't seems so right, but okay.

File: app/assets/javascripts/components/story/CollapsedStory/CollapsedStoryStateActions.js [hunk 13]
```diff
@@ -40,11 +42,20 @@ class CollapsedStoryStateActions extends React.Component {
   }
 
   needConfirmation(action) {
-    return action === storyTransitions.ACCEPT || action === storyTransitions.REJECT;
+    return action === storyTransitions.ACCEPT ||
+      action === storyTransitions.REJECT ||
+      action === storyTransitions.RELEASE;
+  }
+
+  getStoryState(story) {
```
Comment (talyssonoc) [inline app/assets/javascripts/components/story/CollapsedStory/CollapsedStoryStateActions.js]: What do you think about moving this logic to this file https://github.com/Codeminer42/cm42-central/pull/535/files#diff-b59c70dd5bb6cd4a0b5b67e0490f03fdR315 ?

File: app/assets/javascripts/models/beta/story.js [hunk 14]
```diff
@@ -64,7 +65,19 @@ export const getCompletedPoints = story => {
 
 export const isStoryNotEstimated = (storyType, estimate) => storyType === 'feature' && !estimate;
```
Comment (talyssonoc) [inline app/assets/javascripts/models/beta/story.js]: Note for a future refactor (we don't need to do it now): let's just receive the `story` as the only parameter here.
